src/container/srv_target.c

```diff
@@ -1142,146 +1160,278 @@ ds_cont_tgt_epoch_discard_aggregator(crt_rpc_t *source, crt_rpc_t *result,
-		goto pool_child;
+int
+ds_cont_tgt_snapshots_update(uuid_t pool_uuid, uuid_t cont_uuid,
+			     uint64_t *snapshots, int snap_count)
+{
+	struct cont_snap_args	 args;
+
+	uuid_copy(args.pool_uuid, pool_uuid);
+	uuid_copy(args.cont_uuid, cont_uuid);
+	args.snap_count = snap_count;
+	args.snapshots = snapshots;
+	D_DEBUG(DB_TRACE, DF_UUID": refreshing snapshots %d\n",
+		DP_UUID(cont_uuid), snap_count);
+	return dss_thread_collective(cont_snap_update_one, &args, 0);
+}
+
+void
+cont_snapshots_refresh_ult(void *data)
+{
+	struct cont_snap_args	*args = data;
+	struct ds_pool		*pool;
+	int			 rc;
+
+	pool = ds_pool_lookup(args->pool_uuid);
+	if (pool == NULL) {
+		rc = -DER_NO_HDL;
+		goto out;
 	}
+	rc = cont_iv_snapshots_refresh(pool->sp_iv_ns, args->cont_uuid);
+	ds_pool_put(pool);
+out:
+	if (rc != 0)
+		D_WARN(DF_UUID": failed to refresh snapshots IV: rc %d;"
+			" Aggregation may not work correctly\n",
+			DP_UUID(args->cont_uuid), rc);
+	D_FREE(args);
+}
 
-	rc = vos_cont_query(cont_hdl, &cont_info);
-	if (rc != 0) {
-		D_ERROR(DF_CONT": Query container failed: %d\n",
-			DP_CONT(in->tai_pool_uuid, in->tai_cont_uuid), rc);
-		goto cont_close;
+int
+ds_cont_tgt_snapshots_refresh(uuid_t pool_uuid, uuid_t cont_uuid)
+{
+	struct cont_snap_args	*args;
+	int			 rc;
+
+	D_ALLOC(args, sizeof(*args));
+	if (args == NULL)
+		return -DER_NOMEM;
+	uuid_copy(args->pool_uuid, pool_uuid);
+	uuid_copy(args->cont_uuid, cont_uuid);
+	rc = dss_ult_create(cont_snapshots_refresh_ult, args,
+			    DSS_ULT_POOL_SRV, 0, 0, NULL);
+	if (rc != 0)
+		D_FREE(args);
+	return rc;
+}
+
+/** Currently set to DTX_AGG_THRESHOLD_AGE_LOWER **/
+#define DAOS_AGG_THRESHOLD 3600 /** seconds **/
+
+int
+cont_child_aggregate(struct ds_cont_child *cont)
+{
+	daos_epoch_t		 epoch_max;
+	daos_epoch_range_t	 epoch_range;
+	vos_cont_info_t		 cinfo;
+	int			 i;
+	int			 rc;
+
+	if (cont->sc_aggregation_max == 0)
+		return 0;
+
+	rc = vos_cont_query(cont->sc_hdl, &cinfo);
+	if (rc)
+		return rc;
+
+	/** Don't aggregate if no commits after last aggregation **/
+	if (cinfo.ci_hae > cinfo.ci_hce)
+		return 0;
+
+	epoch_max = crt_hlc_get() -
+		    (uint64_t)DAOS_AGG_THRESHOLD * NSEC_PER_SEC;
+
+	if (epoch_max > cinfo.ci_hce)
+		return 0;
+
+	D_ASSERTF(cinfo.ci_hae <= epoch_max,
+		  "Highest aggregated %lu; Max %lu",
+		  cinfo.ci_hae, epoch_max);
+	/*
+	 * Start from the beginning if snapshots were refreshed since
+	 * the last aggregation in order to handle deleted snapshots
+	 */
+	if (cont->sc_aggregated_ts < cont->sc_snapshots_ts) {
+		i = 0;
+		epoch_range.epr_lo = 0UL;
+	} else {
+		/** Find highest snapshot less than last aggregated epoch **/
+		for (i = 0; i < cont->sc_snapshots_nr &&
+			    cont->sc_snapshots[i] < cinfo.ci_hae; ++i)
+		     ;
+		/** Start from HAE if it is outside snapshot boundaries **/
+		if (i == 0 || i == cont->sc_snapshots_nr)
+			epoch_range.epr_lo = cinfo.ci_hae;
+		else
+			epoch_range.epr_lo = cont->sc_snapshots[i-1] + 1;
```

Comment:
This looks not correct to me. The epr_lo should be the nearest snapshot lower than hae (or 0 if there is no such snapshot exists).